File: nodejs-typescript/src/application/IdempotencyRegistry.ts
import type { TransferId } from '../domain/TransferId.js';

export interface IdempotencyRegistry {
  isProcessed(transferId: TransferId): boolean;
  register(transferId: TransferId): void;
}

export class InMemoryIdempotencyRegistry implements IdempotencyRegistry {
  private readonly processed = new Set<string>();

  isProcessed(transferId: TransferId): boolean {
    return this.processed.has(transferId.toString());
  }

  register(transferId: TransferId): void {
    this.processed.add(transferId.toString());
  }
}

File: ruby/lib/mable_bank/application/idempotency_registry.rb
# frozen_string_literal: true

module MableBank
  module Application
    class IdempotencyRegistry
      def seen?(_transfer_id)
        raise NotImplementedError
      end

      def register(_transfer_id)
        raise NotImplementedError
      end
    end

    class InMemoryIdempotencyRegistry < IdempotencyRegistry
      def initialize
        super
        @ids = {}
      end

      def seen?(transfer_id)
        @ids.key?(transfer_id.value)
      end

      def register(transfer_id)
        @ids[transfer_id.value] = true
      end
    end
  end
end
